clusterloader2/pkg/measurement/common/simple/service_creation_latency.go

```diff
@@ -0,0 +1,422 @@
+	pingCheckers  map[string]*pingChecker
+}
+
+// Execute waits until services are ready.
+// Services can be specified by field and/or label selectors.
+// If namespace is not passed by parameter, all-namespace scope is assumed.
+// "start" action starts observation of the services.
+// "waitTillReady" waits until all services are available.
+// "gather" returns service creation latency summary.
+func (s *serviceCreationLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	s.clusterFramework = config.ClusterFramework
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+
+	switch action {
+	case "start":
+		s.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return nil, err
+		}
+		s.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		s.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultServiceCreationLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.start()
+	case "waitTillReady":
+		return nil, s.waitTillReady()
+	case "gather":
+		return s.gather()
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// Dispose cleans up after the measurement.
+func (s *serviceCreationLatencyMeasurement) Dispose() {
+	if s.isRunning {
+		s.isRunning = false
+		close(s.stopCh)
+	}
+	s.queue.Stop()
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for _, pc := range s.pingCheckers {
+		pc.stop()
+	}
+	s.pingCheckers = make(map[string]*pingChecker)
+}
+
+// String returns a string representation of the metric.
+func (*serviceCreationLatencyMeasurement) String() string {
+	return serviceCreationLatencyName
+}
+
+type serviceCreationTimes struct {
+	creation  time.Time
+	assigned  time.Time
+	available time.Time
+}
+
+func (s *serviceCreationLatencyMeasurement) start() error {
+	if s.informer != nil {
+		klog.Infof("%s: wait for controlled pods measurement already running", s)
+		return nil
+	}
+	klog.Infof("%s: starting wait for controlled pods measurement...", s)
+
+	s.isRunning = true
+	s.stopCh = make(chan struct{})
+
+	s.informer = informer.NewInformer(
+		s.clusterFramework.GetClientSets().GetClient(),
+		"services",
+		s.namespace,
+		s.fieldSelector,
+		s.labelSelector,
+		func(oldObj, newObj interface{}) {
+			f := func() {
+				s.handleObject(oldObj, newObj)
+			}
+			s.queue.Add(&f)
+		},
+	)
+	go s.informer.Run(s.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, s.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	return nil
```

Comment:
This whole function looks so boiler-platy - can't we have some "meta-measurement" where we will do this once, instead of separately for every single measurement that needs informer on some resource?

Reply:
Done.